heroic-core/src/main/java/com/spotify/heroic/metric/LocalMetricManager.java

```diff
@@ -389,15 +397,20 @@ private void accept(final Consumer<MetricBackend> op) {
         final AggregationSession session;
         final ResultLimits limits;
         final OptionalLimit groupLimit;
+        final OptionalLimit aggregationLimit;
         final boolean failOnLimits;
 
         @Override
         public void resolved(final Pair<Series, FetchData> result) throws Exception {
             final FetchData f = result.getRight();
-
+            boolean failed = false;
             for (final MetricCollection g : f.getGroups()) {
-                g.updateAggregation(session, result.getLeft().getTags(),
-                    ImmutableSet.of(result.getLeft()));
+                failed = failed ||
+                    failOnLimits && aggregationLimit.isGreater(session::getRetainedMetricsCount);
```

Comment:
This approach would scan through all aggregations and sum up the retained count every time it's fed data.

As an alternative, we could use a system similar to the way that `QuotaWatcher` works. Each aggregation session would be initialized with an `AtomicLong` (wrapped behind an interface) and would be responsible for incrementing this every time it "stored" data. Querying it would only be a matter of reading its value here.

We might also want to consider tying these two systems together, since if the aggregation system "runs out of quota" there is no need to continue reading data from the backend any longer.